// @google-cloud/discoveryengine/build/src/v1beta/index.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserEventServiceClient = exports.SearchServiceClient = exports.SchemaServiceClient = exports.RecommendationServiceClient = exports.DocumentServiceClient = exports.ConversationalSearchServiceClient = exports.CompletionServiceClient = void 0;
var completion_service_client_1 = require("./completion_service_client");
Object.defineProperty(exports, "CompletionServiceClient", { enumerable: true, get: function () { return completion_service_client_1.CompletionServiceClient; } });
var conversational_search_service_client_1 = require("./conversational_search_service_client");
Object.defineProperty(exports, "ConversationalSearchServiceClient", { enumerable: true, get: function () { return conversational_search_service_client_1.ConversationalSearchServiceClient; } });
var document_service_client_1 = require("./document_service_client");
Object.defineProperty(exports, "DocumentServiceClient", { enumerable: true, get: function () { return document_service_client_1.DocumentServiceClient; } });
var recommendation_service_client_1 = require("./recommendation_service_client");
Object.defineProperty(exports, "RecommendationServiceClient", { enumerable: true, get: function () { return recommendation_service_client_1.RecommendationServiceClient; } });
var schema_service_client_1 = require("./schema_service_client");
Object.defineProperty(exports, "SchemaServiceClient", { enumerable: true, get: function () { return schema_service_client_1.SchemaServiceClient; } });
var search_service_client_1 = require("./search_service_client");
Object.defineProperty(exports, "SearchServiceClient", { enumerable: true, get: function () { return search_service_client_1.SearchServiceClient; } });
var user_event_service_client_1 = require("./user_event_service_client");
Object.defineProperty(exports, "UserEventServiceClient", { enumerable: true, get: function () { return user_event_service_client_1.UserEventServiceClient; } });
